src/app/shared/spaces.service.ts

```diff
@@ -1,72 +1,80 @@
     this._current = Observable.merge(
-      contexts.current
-        .map(val => val.space),
-      this.broadcaster.on<Space>('spaceUpdated'));
-    // Create the recent context list
+      this.contexts.current.map(val => val.space),
+      this.broadcaster.on<Space>('spaceUpdated')
+    );
+  }
+
+  private initRecent(): void {
+    // load existing recent spaces from profile.store.recentSpaces
+    this._recent = this.loadRecent().publishReplay(1);
+    this._recent.connect();
 
-    this._recent = Observable.merge(
-      this.broadcaster.on<Space>('spaceChanged'),
-      this.addRecent)
-      // Map from the context being added to an array of recent contexts
-      // The scan operator allows us to access the list of recent contexts and add ours
-      .scan((recent, s) => {
-        // First, check if this context is already in the list
-        // If it is, remove it, so we don't get duplicates
-        for (let i = recent.length - 1; i >= 0; i--) {
-          if (recent[i].id === s.id) {
-            recent.splice(i, 1);
+    // update _recent with newly seen space when spaceChanged is emitted
+    this.broadcaster.on<Space>('spaceChanged')
+      .combineLatest(this._recent)
+      .subscribe((combined: [Space, Space[]]) => {
+        let changedSpace: Space = combined[0];
+        let recentSpaces: Space[] = combined[1];
+        let index: number = this.findRecentIndexById(recentSpaces, changedSpace.id);
+        // if changedSpace exists in _recent
+        if (index !== -1) {
+          // move the space to the front of the list if it is already in _recent
+          if (recentSpaces[0].id !== changedSpace.id) {
+            recentSpaces.splice(index, 1);
+            recentSpaces.unshift(changedSpace);
+            this.saveRecent(recentSpaces);
+          }
+          // otherwise, no operation required; changedSpace is already index 0
+        // if changedSpace is new to the recent spaces
+        } else {
+          recentSpaces.unshift(changedSpace);
+          // trim _recent if required to ensure it is length =< 8
+          if (recentSpaces.length > SpacesService.RECENT_SPACE_LENGTH) {
+            recentSpaces.pop();
           }
+          this.saveRecent(recentSpaces);
         }
-        // Then add this context to the top of the list
-        recent.unshift(s);
-        return recent;
-        // The final value to scan is the initial value, used when the app starts
-      },
-      [])
-      // Finally save the list of recent contexts
-      .do(val => {
-        // Truncate the number of recent contexts to the correct length
-        if (val.length > SpacesService.RECENT_SPACE_LENGTH) {
-          val.splice(
-            SpacesService.RECENT_SPACE_LENGTH,
-            val.length - SpacesService.RECENT_SPACE_LENGTH
-          );
+      }
+    );
+
+    // if a space is deleted, check to see if it should be removed from _recent
+    this.broadcaster.on<Space>('spaceDeleted')
+      .combineLatest(this._recent)
```

Comment:
`combineLatest` will emit an event when either of the observables emit, when `spaceDeleted` is broadcast or when the recent observable emits. In this situation we really only want to do something when there is a `spaceDeleted` broadcast so this can probably be refactored to not run anything when the recent observable emits.

Reply:
I think the `combineLatest` should be replaced by a `broadcaster.on<Space>('spaceDeleted').map()` call, where the map function (flat)maps the single deleted space onto `this._recent.first()`, which in turn is mapped and transformed so as to remove the deleted space. Essentially the logic currently in the subscriber would mostly be moved into the mapping function(s), leaving the subscriber with the sole responsibility of saving the result.